Introducing FreeFlight fallback

delapuente opened the pull request:
Fix #125 
I need some feedback. Look at how it behaves on:

- Desktop supporting WebVR: no fallback is available.
- Desktop not supporting WebVR: movement and orientation fallbacks are available. 
- Mobile: only movement fallback would be available but the absence of keyboard renders it unusable. I'm planning to address this problem in another bug and provide two-finger navigation of the scene.

Use W, A, S, D or arrow keys to move forward, left, backward or right.
Use R, F to increase or decrease elevation.
Use Q, E to roll to the left or to the right.
Use click and drag to yaw and pitch.

cvan commented:
> Desktop not supporting WebVR: movement and orientation fallbacks are available.

oh, btw, in the default scene, the controller hands appear smooshed together at the origin.

![image](https://user-images.githubusercontent.com/203725/37943954-a0fe8a3c-312e-11e8-954b-24903ee1ea6d.png)
![image](https://user-images.githubusercontent.com/203725/37943956-a3703d92-312e-11e8-931f-ff3b7d177a6e.png)

it looks funny 😆but I assume it's a bug/oversight?

delapuente commented on Assets/WebGLTemplates/WebVR/webvr.js:
```diff
@@ -288,24 +288,33 @@
+      if (vrDisplay.capabilities) {
+        if (vrDisplay.capabilities.canPresent) {
+          // Enable button to toggle entering/exiting VR.
+          entervrButton.dataset.enabled = 'true';
```
I made a more in-depth refactor here.

delapuente commented on Assets/WebVR/Scenes/WebVR.unity:
```diff
@@ -1614,6 +1614,24 @@ Prefab:
+  m_Enabled: 1
+  m_EditorHideFlags: 0
+  m_Script: {fileID: 11500000, guid: 683f145ef879be447ba25c1080ac1984, type: 3}
+  m_Name: 
```
It seems to be OK. Unity writes these files. If there was an error, I suppose Unity would complain about it. I don't see any error but if this is a bug, let's report it in a different bug.

delapuente commented on Assets/WebVR/Scenes/WebVR.unity:
```diff
@@ -1614,6 +1614,24 @@ Prefab:
+  m_EditorHideFlags: 0
+  m_Script: {fileID: 11500000, guid: 683f145ef879be447ba25c1080ac1984, type: 3}
+  m_Name: 
+  m_EditorClassIdentifier: 
```
Same as above here.

delapuente commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,134 @@
+
+public class FreeFlightController : MonoBehaviour {
+
+	public float translationSpeed = 7;
```
Why do you think `setters` and `getters` are important here? They are not a pattern we are using in other parts of the code and they tend to add bloat to the code.

delapuente commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,134 @@
+		}
+	}
+
+	public void RequestFallback(string mode="both") {
```
I'm removing this API in favour of calling `onVRCapabilities` from the template. This way, the developer can control what happens in Unity which confers a higher degree of customization and freedom without modifying the template (I think this should be a top priority since our target developers are Unity devs, not web devs, what do you think?).

delapuente commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,134 @@
+
+	void Update () {
+		TrackMouse();
+		if (isTranslationEnabled) {
```
The default value for a `bool` type is `false` but I'll do it explicit.

delapuente commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,134 @@
+		return scale * direction * Time.deltaTime;
+	}
+
+	float QuantifiedDirection(float number, float threshold=0.001f) {
```
These controls are a replacement for the polyfill controls which provided no configuration at all so I would prefer to provide as few customization options as possible. Let's begin by providing a quick replacement class and see what the community needs really are.

delapuente commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,134 @@
+		if (number < -threshold) {
+			return -1;
+		}
+		return 0;
```
It depends if you consider the threshold to be zero. I'm doing it so that 0 means the number is in between `[-threshold, +threshold]`.

cvan replied:
yeah, I did - is this convention?

delapuente commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,134 @@
+
+	float RollMovement() {
+		return RotationMovement(GetDirectionFromAxis(
+			new List<KeyCode> { KeyCode.Q },
```
No, right now. But we can monitor the reaction of the community and see what they prefer. See [this other comment](https://github.com/mozilla/unity-webvr-export/pull/213#discussion_r177357035).

delapuente commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,134 @@
+
+	float ElevationMovement() {
+		return TranslationSpeed(GetDirectionFromAxis(
+			new List<KeyCode> { KeyCode.R },
```
[Same as above](https://github.com/mozilla/unity-webvr-export/pull/213#discussion_r177364191).

delapuente commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,134 @@
+
+	float ForwardMovement() {
+		return TranslationSpeed(GetDirectionFromAxis(
+			new List<KeyCode> { KeyCode.W, KeyCode.UpArrow },
```
Not significant. You can check it without the need of a profiler in the demo scene.

delapuente replied:
I've removed constructing the constant objects again and again. I don't think the `if` alternative would improve the performance in a noticeable way.

delapuente commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,134 @@
+
+	float SideMovement() {
+		return TranslationSpeed(GetDirectionFromAxis(
+			new List<KeyCode> { KeyCode.D, KeyCode.RightArrow },
```
Usually `D` means `stride to the right` in FPS games. I strongly think that a Unity user/game player would find it comfortable and natural.

delapuente replied:
All the keys are now customizable.

delapuente commented:
[What happened to the hands](https://github.com/mozilla/unity-webvr-export/pull/213#issuecomment-376378785) seems like a bug. They should not be displayed unless there are controls.

delapuente commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,134 @@
+	float RollMovement() {
+		return RotationMovement(GetDirectionFromAxis(
+			new List<KeyCode> { KeyCode.Q },
+			new List<KeyCode> { KeyCode.E }
```
Right, this kind of rotation is not very usual and I would prefer to remove it completely following the [recommendation on customization](https://github.com/mozilla/unity-webvr-export/pull/213#discussion_r177364191).

delapuente commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,134 @@
+
+	bool isTranslationEnabled;
+
+	void Update () {
```
[`FixedUpdate` should be used when updating a rigid body](https://unity3d.com/learn/tutorials/topics/scripting/update-and-fixedupdate) inside a physics system, which is not the case.

delapuente commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,134 @@
+	bool isTranslationEnabled;
+
+	void Update () {
+		TrackMouse();
```
According to the documentation:
```
OnMouseDrag is called when the user has clicked on a GUIElement or Collider and is still holding down the mouse.
```
I don't use any `GUIElement` nor `Collider` so I don't think this applies here but we can run an experiment in a different bug.

delapuente commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,134 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class FreeFlightController : MonoBehaviour {
```
Thank you for pointing this to me. It is very interesting but this is intended for [handling mouse interactions where the targets are the `GameObject`s in the scene](https://docs.unity3d.com/Manual/EventSystem.html). It is not a general way to handle mouse draggings.

delapuente commented:
@cvan take another look. The `RequestFallback` API is gone. Now we send the capabilities to Unity so the developer can decide what to do. Actually, the free-flight camera is not special in any way and it can illustrate how to create a non-VR camera for when VR is not available.

I've removed the capability of rolling and added customization options for all the keys moving the camera. Rotation is mouse/touch-screen only.

> currently, in Firefox on Windows the demo scene seems broken in its current state (what with the default camera position and the scenery).

This is tricky and it is related to #184. I would provide a "default-height-for-when-no-vr" component in a different component. I can draft one tomorrow. What do you think?

cvan commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,188 @@
+		return direction * rotationSpeed * Time.deltaTime;
+	}
+
+	float DirectionFromMovement(float number, float threshold=0.001f) {
```
you don't think this should be configurable? probably fine, eh

cvan commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,188 @@
+	}
+
+	void RegisterMouseMovement() {
+		bool stoppedTracking = Input.GetMouseButtonUp(0);
```
IMO "Tracking" is a weird word to use because of VR "positional tracking"

can't think of the best word: `stoppedMouseDrag`, `mouseDragReleased`, `mouseClickReleased`?

cvan commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,188 @@
+	}
+
+	void Update() {
+		if (translationEnabled) {
```
perfect, thanks!

cvan commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,188 @@
+
+public class FreeFlightController : MonoBehaviour {
+
+	[Tooltip("Enable/disable rotation control.")]
```
thanks for moving all the `public` properties and adding the tooltip text labels. they're very helpful (for me definitely anyway)

cvan commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,188 @@
+	public List<KeyCode> moveForwardKeys =
+		new List<KeyCode> { KeyCode.W, KeyCode.UpArrow };
+
+	[Tooltip("Keys to move backwards.")]
```
for consistency: `backwards` -> `backward`

cvan commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,188 @@
+		new List<KeyCode> { KeyCode.W, KeyCode.UpArrow };
+
+	[Tooltip("Keys to move backwards.")]
+	public List<KeyCode> moveBackwardsKeys =
```
`moveBackwardsKeys` -> `moveBackwardKeys`

cvan commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,188 @@
+	public List<KeyCode> moveBackwardsKeys =
+		new List<KeyCode> { KeyCode.S, KeyCode.DownArrow };
+
+	[Tooltip("Keys to stride-right.")]
```
`stride-right` -> `stride to the right`?

cvan commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,188 @@
+	}
+
+	void DisableEverything() {
+		rotationEnabled = false;
```
nit: for consistency with the order in the `Update` method, can you swap the order of these lines?

cvan commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,188 @@
+	/// Enables rotation and translation control for desktop environments.
+	/// For mobile environments, it enables rotation or translation according to
+	/// the device capabilities.
+	void EnableAccordingToPlatform() {
```
correct me if I'm wrong, but these assignments do not respect the user-overridden value if set by the developer from the Settings checkbox options for `rotationEnabled` and `translationEnabled`?

I see this method is then called in these spots:
- https://github.com/mozilla/unity-webvr-export/pull/213/files#diff-5f0267533825be26e8fdcbb935db6cb6R59
- https://github.com/mozilla/unity-webvr-export/pull/213/files#diff-5f0267533825be26e8fdcbb935db6cb6R75

can we check first when the class is constructed (vs. only in the [`Update` method](https://github.com/mozilla/unity-webvr-export/pull/213/files#diff-5f0267533825be26e8fdcbb935db6cb6R78)) that the user hasn't explicitly disabled Rotation or Translation?

delapuente replied:
Good observation but I did on purpose for them being only usable while in the editor. This is because for this class to work as intended we would need kind of a "WebVR emulator" plug-in for the editor that would call `OnVRCapabilities` with user-selected faked capabilities.

The plug-in would also emulate `EnterVR` and `ExitVR` calls. I do not know if this is doable but I do know it is beyond the scope of this bug so I'm going to *clarify that the switches are for testing in editor mode only* and [file a new issue with the enhancement](https://github.com/mozilla/unity-webvr-export/issues/217).

cvan commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,188 @@
+		return 0;
+	}
+
+	bool SomeIsPressed(List<KeyCode> keys) {
```
can you rename this to `AnyKeyIsPressed` or `AnyKeyPressed` (basically just include the word `Key` somewhere in this name)?

cvan commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,188 @@
+		return keys.FindAll(k => Input.GetKey(k)).Count > 0;
+	}
+
+	void RegisterMouseMovement() {
```
perhaps I missed it in a comment, but did you look into using the Unity `MonoBehaviour`'s `onMouseDrag`?

delapuente replied:
According to [the documentation](https://docs.unity3d.com/ScriptReference/MonoBehaviour.OnMouseDrag.html):
> OnMouseDrag is called when the user has clicked on a GUIElement or Collider and is still holding down the mouse.

Similar to what happened with the event systems, this is intended for interacting with things that are _draggable_ and not as a generic way of detecting dragging on the screen.

cvan commented on Assets/WebVR/Scripts/UI/WebUI.cs:
```diff
@@ -0,0 +1,9 @@
+using System.Runtime.InteropServices;
+public class WebUI {
+
+	[DllImport("__Internal")]
```
no biggie, really, just FYI: the `.meta` files use two soft spaces :smile:

cvan commented on ProjectSettings/ProjectVersion.txt:
```diff
@@ -1 +1 @@
-m_EditorVersion: 2017.3.0f3
+m_EditorVersion: 2017.3.1f1
```
want to do in separate commit? or are you fine having this?

delapuente replied:
Sure, I will commit this directly to master.

cvan commented on Assets/WebGLTemplates/WebVR/styles/webvr.css:
```diff
@@ -130,6 +130,7 @@ a:link, a:visited {
   position: absolute;
   top: 0;
   width: 100%;
+  pointer-events: none;
```
I noticed this - thanks for fixing this 👍 :smile:

delapuente commented:
I've managed to add a fixing offset for when there is no position capability or we are in a desktop-like environment. I've also changed the components to the main camera since to fix a couple of edge cases and cleaning some things from bug #159 that interfered with this implementation. @cvan feel free to take another look. Remember I won't be available this week.

cvan commented on Assets/WebVR/Scripts/Input/FreeFlightController.cs:
```diff
@@ -0,0 +1,200 @@
+			mouseMovement.Set(0, 0, 0);
+		}
+		else if (mouseIsDragging) {
+			mouseMovement = Input.mousePosition - lastMousePosition;
```
not sure if it's coming from the `YawMovement` or `PitchMovement`, but the mousedrag (especially at lower regions of the screen) doesn't seem to rotate the camera correctly; it's like a lag or stutter

cvan replied:
filed issue #219

cvan commented on Assets/WebGLTemplates/WebVR/webvr.js:
```diff
@@ -287,26 +300,30 @@
-      }
+      gameInstance.SendMessage(
+        'WebVRCameraSet', 'OnVRCapabilities',
+        JSON.stringify({
```
can handle in #104

cvan commented:
thanks! this is sooo good! 🏅

caseyyee commented:
Nice job @delapuente !
Good one 💯
